### PR TITLE
test for `@clientName` of TCGC and `@encodedName` of compiler

### PR DESCRIPTION
**Pull request by msyyc**
fixes https://github.com/Azure/cadl-ranch/issues/508, https://github.com/Azure/cadl-ranch/issues/509

- These cases are almost same with existing cases for `@projectedName` so devs shall be able to add the test with few cost

### CODE REVIEW

**Review by weidongxu-microsoft — COMMENTED** [review]
Is it a direct copy from https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/projection/projected-name with only decorator change?

**Comment by msyyc** [comment]
> Is it a direct copy from https://github.com/Azure/cadl-ranch/tree/main/packages/cadl-ranch-specs/http/projection/projected-name with only decorator change?

Yeah, almost

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  op json(...JsonClientNameAndEncodedNameModel): NoContentResponse;
+
+  @scenario
+  @scenarioDoc("""
+Testing that we can project the client name in our generated SDKs.
+Your generated SDK should generate ClientClientNameAndEncodedNameModel with one property `clientName` with wire name `defaultName`.
+
+Expected request body:
+```json
+{"defaultName": true}
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:66]: Is this correct?

Maybe double check the docs?

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:66]: i think it is no? property is `defaultName` and it has a `@clientName("clientName")` so over json it is `defaultName`

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+@route("/model")
+@operationGroup
+namespace Model {
+  @clientName("CSModel", "csharp")
+  @clientName("GoModel", "go")
+  @clientName("JavaModel", "java")
+  @clientName("TSModel", "javascript")
+  @clientName("PythonModel", "python")
+  model ModelWithLanguageClientNameAndEncodedName {
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:140]: Not sure this name make sense has it has no `@encodedName`

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    @doc("Pass in true")
+    defaultName: boolean;
+  }
+
+  @clientName("ClientModel")
+  model ModelWithClientClientNameAndEncodedName {
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:146]: Same here no `@encodedName`

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 1 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:66]: Timothee you are correct. Yeah it does not have encodedName. `JsonAndClientClientNameAndEncodedNameModel` is the one.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 2 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:140]: does model name make sense in wire layer ? I feel like no?

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:140]: It doesn’t yeah I am just saying the model names needs to be changed

Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: Got your point, updated.

**Review by qiaozha — APPROVED** [review]
LGTM, we should remove projectedName test cases after it's offically announced deprecated  https://github.com/microsoft/typespec/issues/2832

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+namespace Property {
+  model LanguageClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @clientName("CSName", "csharp")
+    @clientName("GoName", "go")
+    @clientName("JavaName", "java")
+    @clientName("TSName", "javascript")
+    @clientName("python_name", "python")
+    defaultName: boolean;
+  }
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:24]: recommend `LanguageClientNameModel`?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+  model JsonClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:30]: recommend `JsonEncodedNameModel`?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    @clientName("clientName")
+    defaultName: boolean;
+  }
+
+  model JsonAndClientClientNameAndEncodedNameModel {
+    @doc("Pass in true")
+    @clientName("clientName")
+    @encodedName("application/json", "wireName")
+    defaultName: boolean;
+  }
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:43]: similar for them

**Review by tadelesh — COMMENTED** [review]
Can we also add test for client renaming of `interface` or `namespace`?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+Expected query parameter: `default-name="true"`
+
+""")
+@route("/parameter")
+@post
+op parameter(
+  @clientName("clientName")
+  @query
+  `default-name`: string,
```
Comment (tadelesh) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:129]: Can we add test for query/header wire name renaming?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 4 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: Updated.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 5 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: Updated.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 6 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: I changed `ClientClientNameAndEncodedNameModel`  as `ClientNameModel` and  `JsonAndClientClientNameAndEncodedNameModel` as `ClientNameAndJsonEncodedNameModel`. let me know if that makes sense ?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 7 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: added here https://github.com/Azure/cadl-ranch/pull/511/files#diff-64d6117e7c3da950e6cf3f1b07c36975dc89959e0848cb1da15af7a2e8993b20R116-R149

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+```json
+{"wireName": true}
+```
+""")
+  @route("/json-and-client")
+  @post
+  op jsonAndClient(...ClientNameAndJsonEncodedNameModel): NoContentResponse;
+}
+
+// May add query/header project after https://github.com/microsoft/cadl/issues/1487
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:102]: nit, remove?

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+""")
+@route("/header")
+@post
+op header(
+  @clientName("clientName")
+  @encodedName("application/json", "header-name")
+  @header
+  defaultName: string,
+): NoContentResponse;
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:148]: nit, may put a header in response as well?

**Review by weidongxu-microsoft — APPROVED** [review]

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 7 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: It seems that `@encodedName` can not be apply to query or header parameter. See [typespec playground link](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwpAc2VydmljZSh7CiAgdGl0bGU6ICJXaWRnZXQgU8YbIiwKICB2ZXJzaW9uOiAiMS4wLjAiLAp9KQpuYW1lc3BhY2UgRGVtb8cvOwoKQHJvdXRlKCIvcXVlcnkiKQpAcG9zdApvcCDFESgKICBAZW5jb2RlZE5hbWUoImFwcGxpY2F0aW9uL2pzb24iLCAixSstxGMiKcQxxRAKICBkZWZhdWx0xDk6IHN0cmluZywKKTogTm9Db250ZW50UmVzcG9uc%2B0AiWhlYWRlcuwAisYS%2FwCL5wCLxizrAIzGEf8Aje8AjQ%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+Your generated SDK should generate an operation `parameter` with a single parameter called `clientName`.
+
+Expected query parameter: `query-name="true"`
+
+""")
+@route("/parameter")
+@post
+op parameter(
+  @clientName("clientName")
+  @encodedName("application/json", "query-name")
```
Comment (chunyu3) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:127]: It is a little confuse to define a query parameter for content-type `application/json`. content-type is for body parameter.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+Your generated SDK should generate an operation header `parameter` with a single parameter called `clientName`.
+
+Expected header parameter: `header-name="true"`
+
+""")
+@route("/header")
+@post
+op header(
+  @clientName("clientName")
+  @encodedName("application/json", "header-name")
```
Comment (chunyu3) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:144]: same as above

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+  @query
+  defaultName: string,
+): NoContentResponse;
+
+@scenario
+@scenarioDoc("""
+Testing that we can project a header name.
+Your generated SDK should generate an operation header `parameter` with a single parameter called `clientName`.
+
+Expected header parameter: `header-name="true"`
```
Comment (chunyu3) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:137]: need to add expected response here. In the expected response, the header is `header-name`

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 10 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:127]: Thanks, I also found currently `@encodedName` is not working with the query and header parameter. left a comment here https://github.com/Azure/cadl-ranch/pull/511#discussion_r1495377968

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 7 above)
Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: Yeah it doesn’t apply to query params as those are not json things. Not sure if we wanted to support that what is the right media type for it if any. To change the name of header, query you can use the param of the corresponding decorators

Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: I have removed the encodedName for query and header parameter and header in response.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 10 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: updated.

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 12 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp]: Added. Thanks for the suggestion :)

File: packages/cadl-ranch-api/src/types.ts [hunk 13]
```diff
@@ -42,9 +42,8 @@ export type KeyedMockRequestHandler<T extends string = string> = (
 
 export type HttpMethod = "get" | "post" | "put" | "patch" | "delete" | "head" | "options";
 
-export type MockApiForHandler<Handler extends MockRequestHandler> = Handler extends KeyedMockRequestHandler<infer K>
-  ? KeyedMockApi<K>
-  : MockApi;
+export type MockApiForHandler<Handler extends MockRequestHandler> =
+  Handler extends KeyedMockRequestHandler<infer K> ? KeyedMockApi<K> : MockApi;
```
Comment (qiaozha) [inline packages/cadl-ranch-api/src/types.ts:46]:
```suggestion
export type MockApiForHandler<Handler extends MockRequestHandler> = Handler extends KeyedMockRequestHandler<infer K>
  ? KeyedMockApi<K>
  : MockApi;
```

File: packages/cadl-ranch-expect/src/decorators.ts [hunk 14]
```diff
@@ -127,8 +127,8 @@ function getRouteSegments(program: Program, target: Operation | Interface | Name
       return target.interface
         ? [...getRouteSegments(program, target.interface), ...seg]
         : target.namespace
-        ? [...getRouteSegments(program, target.namespace), ...seg]
-        : seg;
+          ? [...getRouteSegments(program, target.namespace), ...seg]
+          : seg;
```
Comment (qiaozha) [inline packages/cadl-ranch-expect/src/decorators.ts:131]:
```suggestion
        ? [...getRouteSegments(program, target.namespace), ...seg]
        : seg;
```

File: packages/cadl-ranch/src/server/server.ts [hunk 15]
```diff
@@ -17,8 +17,8 @@ const errorHandler: ErrorRequestHandler = (err, _req, res, _next) => {
   const errResponse = err.toJSON
     ? err.toJSON()
     : err instanceof Error
-    ? { name: err.name, message: err.message, stack: err.stack }
-    : err;
+      ? { name: err.name, message: err.message, stack: err.stack }
+      : err;
```
Comment (qiaozha) [inline packages/cadl-ranch/src/server/server.ts:21]:
```suggestion
    ? { name: err.name, message: err.message, stack: err.stack }
    : err;
```

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+@scenarioDoc("""
+Testing that we can project a header name.
+Your generated SDK should generate an operation header `parameter` with a single parameter called `clientName`.
+
+Expected header parameter: `default-name="true"`
+Expected response header: `default-name="true"`
+""")
+@route("/header")
+@post
+op header(
```
Comment (chunyu3) [inline packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:140]: shall we also add a scenario to decorate the @clientName on a query?

**Comment by chunyu3** [comment]
others LGTM

**Review by tadelesh — APPROVED** [review]

File: packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp (hunk 16 above)
Comment (qiaozha) [reply packages/cadl-ranch-specs/http/projection/client-encoded-name/main.tsp:140]: We have it here https://github.com/Azure/cadl-ranch/pull/511/files#diff-64d6117e7c3da950e6cf3f1b07c36975dc89959e0848cb1da15af7a2e8993b20R114-R128

**Pull request merged**